tests/integration/pilot/cni/cni_race_test.go

```diff
@@ -0,0 +1,158 @@
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	istioKube "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/cluster"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/istioctl"
+	"istio.io/istio/pkg/test/framework/label"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/shell"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/integration/security/util"
+	"istio.io/istio/tools/istio-iptables/pkg/constants"
+)
+
+var (
+	ist  istio.Instance
+	apps = &util.EchoDeployments{}
+)
+
+func TestMain(m *testing.M) {
+	framework.NewSuite(m).
+		Label(label.CustomSetup).
+		Setup(istio.Setup(&ist, nil)).
+		Setup(func(t resource.Context) error {
+			return util.SetupApps(t, ist, apps, false)
+		}).
+		Run()
+}
+
+func TestCNIRaceRepair(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.cni.race-condition-repair").
+		Run(func(t framework.TestContext) {
+			if !ist.Settings().EnableCNI {
+				t.Skip("CNI race condition mitigation is only tested when CNI is enabled.")
+			}
+			cluster := t.Clusters().Default()
+			// To begin with, delete CNI Daemonset to simulate a CNI race condition.
+			deleteCNIDaemonset(t, cluster)
+
+			// Rollout restart instances in namespace 1, and wait for a broken instance.
+			rolloutCmd := fmt.Sprintf("kubectl rollout restart deployment -n %s", apps.Namespace1.Name())
+			if _, err := shell.Execute(true, rolloutCmd); err != nil {
+				t.Fatalf("failed to rollout restart deployments %v", err)
+			}
+			waitForBrokenPodOrFail(t, cluster)
+
+			// Now bring back CNI Daemonset, and pod in namespace 1 should be repaired.
+			if err := deployCNIDaemonset(t, cluster); err != nil {
+				t.Fatalf("failed to deploy CNI Dameonset %v", err)
+			}
+			waitForRepairOrFail(t, cluster)
+		})
+}
+
+func deleteCNIDaemonset(ctx framework.TestContext, c cluster.Cluster) {
+	if err := c.(istioKube.ExtendedClient).
+		Kube().AppsV1().DaemonSets("kube-system").
+		Delete(context.Background(), "istio-cni-node", metav1.DeleteOptions{}); err != nil {
+		ctx.Fatalf("failed to delete CNI Daemonset %v", err)
+	}
+
+	// Wait until the CNI Daemonset pod cannot be fetched anymore
+	retry.UntilSuccessOrFail(ctx, func() error {
+		scopes.Framework.Infof("Checking if CNI Daemonset pod is deleted...")
+		pods, err := c.PodsForSelector(context.TODO(), "kube-system", "k8s-app=istio-cni-node")
+		if err != nil {
+			return err
+		}
+		if len(pods.Items) > 0 {
+			return errors.New("CNI Daemonset pod still exists after deletion")
+		}
+		return nil
+	}, retry.Delay(1*time.Second), retry.Timeout(80*time.Second))
+}
+
+func deployCNIDaemonset(ctx framework.TestContext, c cluster.Cluster) error {
+	args := []string{
+		"install", "-f", filepath.Join(env.IstioSrc,
+			"tests/integration/pilot/cni/testdata/cni.yaml"),
```

Comment:
can we save the old yaml and re-apply it? There may be customizations done outside the test framework?

Reply:
Done.